ros_gz_bridge/src/convert/marine_acoustic_msgs.cpp

```diff
@@ -0,0 +1,165 @@
+
+    // rssi not supported in ROS.
+    // nsd not supported in ROS.
+
+    beam->set_locked(true);
+  }
+
+  // Velocity.
+  gz_msg.mutable_velocity()->set_reference(
+    gz::msgs::DVLKinematicEstimate::DVL_REFERENCE_SHIP);
+  convert_ros_to_gz(ros_msg.velocity, (*gz_msg.mutable_velocity()->mutable_mean()));
+  for (auto i = 0; i < 9; ++i) {
+    gz_msg.mutable_velocity()->add_covariance(ros_msg.velocity_covar[i]);
+  }
+
+  // Target.
+  if (ros_msg.velocity_mode == marine_acoustic_msgs::msg::Dvl::DVL_MODE_BOTTOM) {
+    gz_msg.mutable_target()->set_type(gz::msgs::DVLTrackingTarget::DVL_TARGET_BOTTOM);
+  } else if (ros_msg.velocity_mode == marine_acoustic_msgs::msg::Dvl::DVL_MODE_WATER) {
+    gz_msg.mutable_target()->set_type(gz::msgs::DVLTrackingTarget::DVL_TARGET_WATER_MASS);
+  } else {
+    gz_msg.mutable_target()->set_type(gz::msgs::DVLTrackingTarget::DVL_TARGET_UNSPECIFIED);
+  }
+
+  // Range and position Unavailable.
+
+  gz_msg.set_status(0);
+}
+
+template<>
+void
+convert_gz_to_ros(
+  const gz::msgs::DVLVelocityTracking & gz_msg,
+  marine_acoustic_msgs::msg::Dvl & ros_msg)
+{
+  convert_gz_to_ros(gz_msg.header(), ros_msg.header);
+
+  if (gz_msg.target().type() == gz::msgs::DVLTrackingTarget::DVL_TARGET_BOTTOM) {
+    ros_msg.velocity_mode = marine_acoustic_msgs::msg::Dvl::DVL_MODE_BOTTOM;
+  } else if (gz_msg.target().type() == gz::msgs::DVLTrackingTarget::DVL_TARGET_WATER_MASS) {
+    ros_msg.velocity_mode = marine_acoustic_msgs::msg::Dvl::DVL_MODE_WATER;
+  }
+
+  if (gz_msg.type() == gz::msgs::DVLVelocityTracking::DVL_TYPE_PISTON) {
+    ros_msg.dvl_type = marine_acoustic_msgs::msg::Dvl::DVL_TYPE_PISTON;
+  } else if (gz_msg.type() == gz::msgs::DVLVelocityTracking::DVL_TYPE_PHASED_ARRAY) {
+    ros_msg.dvl_type = marine_acoustic_msgs::msg::Dvl::DVL_TYPE_PHASED_ARRAY;
+  }
+
+  convert_gz_to_ros(gz_msg.velocity().mean(), ros_msg.velocity);
+
+  for (auto i = 0; i < 9; ++i) {
+    if (gz_msg.velocity().covariance_size() > i) {
+      ros_msg.velocity_covar[i] = gz_msg.velocity().covariance()[i];
+    } else {
+      ros_msg.velocity_covar[i] = -1;
+    }
+  }
+
+  ros_msg.altitude = -1;
+  ros_msg.course_gnd = std::atan2(ros_msg.velocity.y, ros_msg.velocity.x);
+  ros_msg.speed_gnd = std::sqrt(ros_msg.velocity.x * ros_msg.velocity.x + ros_msg.velocity.y *
+      ros_msg.velocity.y);
+
+  // Unsupported in Gazebo.
+  ros_msg.sound_speed = -1;
+
+  ros_msg.beam_ranges_valid = false;
+  ros_msg.beam_velocities_valid = false;
+
+  // Crop num beams if needed.
+  uint8_t numGoodBeams = 0u;
+  auto numBeams = std::min(gz_msg.beams_size(), 4);
+
+  for (auto i = 0; i < numBeams; ++i) {
+    if (!gz_msg.beams()[i].locked()) {
+      continue;
+    }
+
+    ros_msg.beam_ranges_valid = true;
+    ros_msg.beam_velocities_valid = true;
+
+    // beam_unit_vec is unsupported.
+    ros_msg.beam_unit_vec[numGoodBeams].x = -1;
+    ros_msg.beam_unit_vec[numGoodBeams].y = -1;
+    ros_msg.beam_unit_vec[numGoodBeams].z = -1;
```

Comment:
@caguero why not use the normalized mean velocity vector? I don't know if the frames match but we are already assuming they do for the ros to gz conversion.